### PR TITLE
Add basic workflow for checkcode

### LINKED ISSUE #2: Use `Arc<dyn Trait>` instead of `Arc<Box<dyn Trait>>`
It's [redundant allocation](https://rust-lang.github.io/rust-clippy/master/index.html#redundant_allocation).

Seems you don't use [Clippy](https://github.com/rust-lang/rust-clippy), it's a very useful linter.

### PR DESCRIPTION
**Pull request by PsiACE**
good luck

in workflow:

- [x] cargo build/test on macos/ubuntu
- [x] cargo clippy/fmt on ubuntu

other changes:

- [x] make clippy happy
  - [x] re-org modules
  - [x] Arc<Box<dyn Trait>> -> Arc<dyn Trait>
  - [x] ...
- [x] pass all test
- [x] fix docs

fixes #2

### CODE REVIEW

**Comment by PsiACE** [comment]
i will fix it later

**Comment by huachaohuang** [comment]
Wow, you sent the first PR. Thanks!

File: engula/src/format/iterator.rs [hunk 1]
```diff
@@ -24,15 +24,11 @@ impl<'a> PartialEq for Version<'a> {
...
-        if ord == Ordering::Equal {
-            if self.0 < other.0 {
-                ord = Ordering::Greater;
-            } else if self.0 > other.0 {
-                ord = Ordering::Less;
-            }
+        match self.1.cmp(other.1) {
+            Ordering::Greater => Ordering::Greater,
+            Ordering::Less => Ordering::Less,
+            Ordering::Equal => other.0.cmp(&self.0),
```
Comment (huachaohuang) [inline engula/src/format/iterator.rs:30]: Nice :)

**Comment by PsiACE** [comment]
see workflow result here: https://github.com/PsiACE/engula/actions/runs/1194972163

Good luck, Rustaceans, the friendly community blessed us.

My wechat is PsiACE, keep in touch. cc @huachaohuang

**Review by huachaohuang — APPROVED** [review]
Well done, you fixed all the code, LGTM

**Comment by huachaohuang** [comment]
@PsiACE thanks again, you are our first contributor! We will send you some souvenirs once we are ready :)

**Pull request merged**